### PR TITLE
[ntlm]: use the hash algorithm as mandated by rfc5929

### PR DESCRIPTION
**Pull request by akarl10**
This is for making ntlm (and probably Negotiate/Kerberos too) with gateways that use sha384 or sha512 certificate hashes.

### CODE REVIEW

File: libfreerdp/crypto/crypto.c [hunk 1]
```diff
@@ -289,6 +289,30 @@ BYTE* crypto_cert_hash(X509* xcert, const char* hash, UINT32* length)
 	return fp;
 }
 
+char* cryto_cert_hash_use_signature_alg(X509* xcert)
```
Comment (akallabeth) [inline libfreerdp/crypto/crypto.c:292]: @akarl10 ok, if you want to use it this way better use a `char* crypto_cert_get_signature_alg(X509* xcert, char* buffer, size_t buffersize)` function signature

**Review by akallabeth — APPROVED** [review]
Basically LGTM, I´d just rename the function and use a supplied buffer here.

**Comment by akallabeth** [comment]
@freerdp-bot test this nice pr please!

File: libfreerdp/crypto/tls.c [hunk 2]
```diff
@@ -616,7 +616,21 @@ static SecPkgContext_Bindings* tls_get_channel_bindings(X509* cert)
 	SEC_CHANNEL_BINDINGS* ChannelBindings;
 	SecPkgContext_Bindings* ContextBindings;
 	const size_t PrefixLength = strnlen(TLS_SERVER_END_POINT, ARRAYSIZE(TLS_SERVER_END_POINT));
-	BYTE* CertificateHash = crypto_cert_hash(cert, "sha256", &CertificateHashLength);
+
```
Comment (akallabeth) [inline libfreerdp/crypto/tls.c:619]: Maybe add a link to the RFX in a comment here that we know what is going on in the future

File: libfreerdp/crypto/crypto.c (hunk 1 above)
Comment (akallabeth) [reply libfreerdp/crypto/crypto.c:292]: another option would be to return a `WINPR_MD_TYPE`

Comment (akarl10) [reply libfreerdp/crypto/crypto.c:292]: I think maybe the second option would be better. The buffer could be hidden inside the function and at the end of the day WINPR_MD_TYPE is what we need. `winpr_md_type_to_string` returns a pointer to a const char*, so we don't need to bother about the allocation

File: libfreerdp/crypto/tls.c (hunk 2 above)
Comment (akarl10) [reply libfreerdp/crypto/tls.c:619]: definitly makes sense

File: libfreerdp/crypto/crypto.c [hunk 3]
```diff
@@ -289,6 +289,30 @@ BYTE* crypto_cert_hash(X509* xcert, const char* hash, UINT32* length)
...
+	hashName = calloc(50, sizeof(char));
+	if (hashName == NULL)
+		return NULL;
+
+	X509_ALGOR_get0(&sigAlgObj, NULL, NULL, tsig_alg);
+	OBJ_obj2txt(hashName, 50, sigAlgObj, 0);
+
+	/* OBJ_obj2txt returns hashWithEncryption, ex. sha256WithRSAEncrpytion;
+	 * we need only the part before With */
+	char* w = strstr(hashName, "With");
```
Comment (akallabeth) [inline libfreerdp/crypto/crypto.c:309]: this check is not sufficient, it only covers `RSA` certificates.
Might be better to use `X509_get_signature_nid` and do a big switch for all (currently) supported algorighm combinations (maybe there is something simpler?)

File: libfreerdp/crypto/crypto.c (hunk 1 above)
Comment (akallabeth) [reply libfreerdp/crypto/crypto.c:292]: good point

Comment (akarl10) [reply libfreerdp/crypto/crypto.c:292]: is it allowed to include something from winpr in inlcude/freerdp/crypto? I get `crypto.h:60:21: error: unknown type name ‘WINPR_MD_TYPE’`

Comment (akallabeth) [reply libfreerdp/crypto/crypto.c:292]: @akarl10 yes, just not the other way around

**Comment by akallabeth** [comment]
@akarl10 https://stackoverflow.com/questions/36076998/how-to-get-the-type-of-signature-hash-algorithm-from-certificate might be of interest, that way you get the hash algorithm `NID` directly

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test again please

**Comment by akallabeth** [comment]
@freerdp-bot test please

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test again please

**Comment by akallabeth** [comment]
@freerdp-bot test please

**Comment by akallabeth** [comment]
@freerdp-bot test again please

**Pull request merged**